pageserver/src/page_service.rs

```diff
@@ -899,165 +995,383 @@ impl PageServerHandler {
+
+        let cancel = self.cancel.clone();
+
+        //
+        // Create Reading future.
+        //
+
+        let (requests_tx, mut requests_rx) = tokio::sync::mpsc::channel(1);
+        let read_messages = {
+            let cancel = self.cancel.child_token();
+            let ctx = ctx.attached_child();
+            async move {
+                scopeguard::defer! {
+                    debug!("exiting");
+                }
+                let mut pgb_reader = pgb_reader;
+                loop {
+                    let msg = Self::pagestream_read_message(
+                        &mut pgb_reader,
+                        tenant_id,
+                        timeline_id,
+                        &mut timeline_handles,
+                        &cancel,
+                        &ctx,
+                        request_span.clone(),
+                    )
+                    .await?;
+                    let msg = match msg {
+                        Some(msg) => msg,
+                        None => {
+                            debug!("pagestream subprotocol end observed");
+                            break;
+                        }
+                    };
+                    match requests_tx.send(msg).await {
+                        Ok(()) => {}
+                        Err(tokio::sync::mpsc::error::SendError(_)) => {
+                            debug!("downstream is gone");
+                            break;
+                        }
                     }
-                };
+                }
+                Ok((pgb_reader, timeline_handles))
+            }
+        }
+        .instrument(tracing::info_span!("read_messages"));
+
+        //
+        // Create Batching future.
+        //
 
-                // Map handler result to protocol behavior.
-                // Some handler errors cause exit from pagestream protocol.
-                // Other handler errors are sent back as an error message and we stay in pagestream protocol.
-                for handler_result in handler_results {
-                    let response_msg = match handler_result {
-                        Err(e) => match &e {
-                            PageStreamError::Shutdown => {
-                                // If we fail to fulfil a request during shutdown, which may be _because_ of
-                                // shutdown, then do not send the error to the client.  Instead just drop the
-                                // connection.
-                                span.in_scope(|| info!("dropping connection due to shutdown"));
-                                return Err(QueryError::Shutdown);
+        enum BatchState {
+            Building(Option<Box<BatchedFeMessage>>),
+            ReadMessagesEnded(Option<Box<BatchedFeMessage>>),
+            ExecutorEnded,
+        }
+        let (batch_tx, mut batch_rx) = tokio::sync::watch::channel(Arc::new(
+            std::sync::Mutex::new(BatchState::Building(None)),
```

Comment:
`watch::channel` already wraps the value in a read/write lock. Can we skip this Mutex?

Reply:
The `watch::Receiver` doesn't allow mutable access, so, yes, this Mutex is necessary in the current constellation.

I asked in #rust for a better primitive: https://neondb.slack.com/archives/C0277TKAJCA/p1732277778048099

Reply:
implemented in https://github.com/neondatabase/neon/pull/9851/commits/9bf2618c45f4d13400ab915a49b05160c18e7f2f and adopted in https://github.com/neondatabase/neon/pull/9851/commits/a23abb2cc0c88abf2dc615e37ab3ddc56d953c73